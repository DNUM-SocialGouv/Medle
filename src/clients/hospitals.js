import fetch from "isomorphic-unfetch"

import { API_URL, HOSPITALS_ENDPOINT } from "../config"
import { handleAPIResponse } from "../utils/errors"
import { METHOD_DELETE, METHOD_POST, METHOD_PUT } from "../utils/http"
import { isEmpty } from "../utils/misc"

export const searchHospitalsFuzzy = async ({ search, headers }) => {
  const bonus = search ? `?fuzzy=${search}` : ""

  const response = await fetch(`${API_URL}${HOSPITALS_ENDPOINT}${bonus}`, { headers })
  const hospitals = await handleAPIResponse(response)
  return isEmpty(hospitals) ? [] : hospitals
}

export const findHospital = async ({ id, headers }) => {
  const response = await fetch(`${API_URL}${HOSPITALS_ENDPOINT}/${id}`, { headers })
  return handleAPIResponse(response)
}

export const findAllHospitals = async () => {
  const response = await fetch(`${API_URL}${HOSPITALS_ENDPOINT}`)
  return handleAPIResponse(response)
}

export const createHospital = async ({ hospital, headers }) => {
  const response = await fetch(`${API_URL}${HOSPITALS_ENDPOINT}`, {
    method: METHOD_POST,
    headers: { ...headers, "Content-Type": "application/json" },
    body: JSON.stringify(hospital),
  })
  return handleAPIResponse(response)
}

export const updateHospital = async ({ hospital, headers }) => {
  const response = await fetch(`${API_URL}${HOSPITALS_ENDPOINT}/${hospital.id}`, {
    method: METHOD_PUT,
    headers: { ...headers, "Content-Type": "application/json" },
    body: JSON.stringify(hospital),
  })
  return handleAPIResponse(response)
}

export const deleteHospital = async ({ id, headers }) => {
  const response = await fetch(`${API_URL}${HOSPITALS_ENDPOINT}/${id}`, { method: METHOD_DELETE, headers })
  return handleAPIResponse(response)
}
