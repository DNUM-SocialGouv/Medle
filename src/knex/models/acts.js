import moment from "moment"

export const buildActsFromDB = acts =>
   acts.map(elt => ({
      id: elt.id,
      createdAt: elt.created_at,
      updatedAt: elt.updated_at,
      deletedAt: elt.deleted_at,
      casesFk: elt.cases_fk,
      pvNumber: elt.pv_number,
      examinationDate: elt.examination_date,
      asker: elt.asker,
      addedBy: elt.added_by,
      data: elt.data,
   }))

// from js to db
export const buildActFromJSON = data => {
   const res = { extra_data: {} }

   const mainKeys = {
      id: "id",
      internalNumber: "internal_number",
      pvNumber: "pv_number",
      examinationDate: "examination_date",
      profile: "profile",
      askerId: "asker_id",
      addedBy: "added_by",
      hospitalId: "hospital_id",
   }

   Object.keys(data).forEach(key => {
      if (mainKeys[key]) {
         res[mainKeys[key]] = data[key]
      } else {
         res.extra_data[key] = data[key]
      }
   })

   return res
}

// from db to js
export const buildActFromDB = data => ({
   id: data.id,
   internalNumber: data.internal_number,
   pvNumber: data.pv_number,
   examinationDate: moment(data.examination_date).format("YYYY-MM-DD"),
   profile: data.profile,
   asker: { id: data.asker_id, name: data.asker_name },
   user: {
      id: data.added_by,
      firstName: data.user_first_name,
      lastName: data.user_last_name,
      email: data.user_email,
   },
   hospital: { id: data.hospital_id, name: data.hospital_name },
   ...data.extra_data,
})
