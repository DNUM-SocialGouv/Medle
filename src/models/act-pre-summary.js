import moment from "moment"

import { ISO_DATE } from "../utils/date"

// from Knex entity to model (JS) entity
export const transform = (knexData) => {
    return !knexData
        ? null
        : {
            examinationDate: moment(knexData.examination_date).format(ISO_DATE),
            hospital: { id: knexData.hospital_id, name: knexData.hospital_name },
            examined: knexData.examined,
            actType: knexData.actType,
            violenceType: knexData.violenceType,
            location: knexData.location,
            age: knexData.age,
            actDuration: knexData.actDuration,
            category: knexData.category,
            ponderation: knexData.ponderation,
            id: knexData.id,
            user: {
                email: knexData.user_email,
                firstName: knexData.user_first_name,
                id: knexData.added_by,
                lastName: knexData.user_last_name,
            },
        }
}

export const transformAll = (list) => list.map((knexData) => transform(knexData))

// from model (JS) entity to Knex entity
export const untransform = (model) => {
    const knexData = { extra_data: {} }

    const mainKeys = {
        id: "id",
        examinationDate: "examination_date",
        category: "category",
        examined: "examined",
        actType: "act_type",
        violenceType: "violence_type",
        location: "location",
        age: "age",
        actDuration: "act_duration",
        ponderation: "ponderation",
        hospitalId: "hospital_id",
        addedBy: "added_by",
    }

    Object.keys(model).forEach((key) => {
        if (mainKeys[key]) {
            knexData[mainKeys[key]] = model[key]
        } else {
            knexData.extra_data[key] = model[key]
        }
    })

    return knexData
}
