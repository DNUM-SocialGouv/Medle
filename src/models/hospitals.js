import * as yup from "yup"

import * as common from "./common"

// Mapping keys from JS model to DB model
const JStoDBKeys = {
  id: "id",
  finesseNumber: "finesse_number",
  name: "name",
  addr1: "addr1",
  addr2: "addr2",
  town: "town",
  depCode: "dep_code",
  postalCode: "postal_code",
}

const schema = yup.object().shape({
  id: yup.number().positive().integer().nullable(),
  finesseNumber: yup.string(),
  name: yup.string(),
  category: yup.string(),
  addr1: yup.string(),
  addr2: yup.string(),
  town: yup.string(),
  depCode: yup.string().matches(/[0-9]{2,3}/),
  postalCode: yup.string().matches(/^$|[0-9]{5}/),
})

export const { transform, transformAll, untransform, validate } = common.build({ JStoDBKeys, schema })
