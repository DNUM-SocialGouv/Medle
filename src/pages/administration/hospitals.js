import React from "react"

const HospitalPage = () => {
   return null
}

export default HospitalPage
