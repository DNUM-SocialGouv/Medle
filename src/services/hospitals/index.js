import { create } from "./create"
import { del } from "./delete"
import { find, findList } from "./find"
import { hospitalsOfUser, search } from "./search"
import { update } from "./update"

export { create, del, find, findList, hospitalsOfUser, search, update }
