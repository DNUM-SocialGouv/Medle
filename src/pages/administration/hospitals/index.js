import AddIcon from "@material-ui/icons/Add"
import Link from "next/link"
import { PropTypes } from "prop-types"
import React, { useState } from "react"
import { Alert, Col, Container, Form, FormGroup, Input, Spinner, Table } from "reactstrap"

import { searchHospitalsFuzzy } from "../../../clients/hospitals"
import { SearchButton } from "../../../components/form/SearchButton"
import Layout from "../../../components/Layout"
import { Title1 } from "../../../components/StyledComponents"
import { buildAuthHeaders, redirectIfUnauthorized, withAuthentication } from "../../../utils/auth"
import { logError } from "../../../utils/logger"
import { ADMIN } from "../../../utils/roles"

const AdminHospitalPage = ({ hospitals: initialHospitals, currentUser }) => {
  const [hospitals, setHospitals] = useState(initialHospitals)
  const [search, setSearch] = useState("")
  const [error, setError] = useState("")
  const [loading, setLoading] = useState(false)

  const onChange = (e) => {
    setSearch(e.target.value)
  }

  const onSubmit = async (e) => {
    e.preventDefault()
    setLoading(true)
    try {
      setHospitals(await searchHospitalsFuzzy({ search }))
    } catch (error) {
      setError("Erreur serveur.")
    } finally {
      setLoading(false)
    }
  }

  return (
    <Layout page="hospitals" currentUser={currentUser} admin={true}>
      <Container
        style={{ maxWidth: 980, minWidth: 740 }}
        className="mt-5 mb-5 d-flex justify-content-between align-items-baseline"
      >
        <Title1 className="">{"Administration des établissements"}</Title1>
        <Link href="/administration/hospitals/[id]" as={`/administration/hospitals/new`}>
          <a>
            <SearchButton className="btn-outline-primary">
              <AddIcon />
              &nbsp; Ajouter
            </SearchButton>
          </a>
        </Link>
      </Container>

      <Container style={{ maxWidth: 980, minWidth: 740 }}>
        <Form onSubmit={onSubmit}>
          <FormGroup row inline className="mb-4 justify-content-center">
            <Col className="flex-grow-1">
              <Input
                type="text"
                name="es"
                id="es"
                placeholder="Rechercher un établissement par son nom, etc."
                value={search}
                onChange={onChange}
                autoComplete="off"
              />
            </Col>
            <Col className="flex-grow-0">
              <SearchButton className="btn-primary" disabled={loading} onClick={onSubmit}>
                {loading ? <Spinner size="sm" color="light" data-testid="loading" /> : "Chercher"}
              </SearchButton>
            </Col>
          </FormGroup>
        </Form>
        {error && (
          <Alert color="danger" className="mb-4">
            {error}
          </Alert>
        )}
        {!error && !hospitals.length && <div className="text-center">{"Aucun établissement trouvé."}</div>}

        {!error && !!hospitals.length && (
          <>
            <Table responsive className="table-hover">
              <thead>
                <tr className="table-light">
                  <th>Nom</th>
                  <th>N° Finess</th>
                  <th>Ville</th>
                  <th>Code postal</th>
                </tr>
              </thead>
              <tbody>
                {hospitals.map((hospital) => (
                  <Link
                    key={hospital.id}
                    href="/administration/hospitals/[hid]"
                    as={`/administration/hospitals/${hospital.id}`}
                  >
                    <tr>
                      <td>
                        <b>{`${hospital.name}`}</b>
                      </td>
                      <td>{hospital.finesseNumber}</td>
                      <td>{hospital.town}</td>
                      <td>{hospital.postalCode}</td>
                      <td>
                        <Link href="/administration/hospitals/[hid]" as={`/administration/hospitals/${hospital.id}`}>
                          <a className="text-decoration-none">Voir</a>
                        </Link>
                      </td>
                    </tr>
                  </Link>
                ))}
              </tbody>
            </Table>
          </>
        )}
      </Container>
    </Layout>
  )
}

AdminHospitalPage.getInitialProps = async (ctx) => {
  const headers = buildAuthHeaders(ctx)

  try {
    const hospitals = await searchHospitalsFuzzy({ headers })
    return { hospitals }
  } catch (error) {
    logError("APP error", error)

    redirectIfUnauthorized(error, ctx)
  }
  return {}
}

AdminHospitalPage.propTypes = {
  hospitals: PropTypes.array,
  currentUser: PropTypes.object.isRequired,
}

export default withAuthentication(AdminHospitalPage, ADMIN)
