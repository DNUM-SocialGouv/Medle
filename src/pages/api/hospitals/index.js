import Cors from "micro-cors"

import { sendAPIError, sendMethodNotAllowedError } from "../../../services/errorHelpers"
import { create, search } from "../../../services/hospitals"
import { checkIsSuperAdmin, checkValidUserWithPrivilege } from "../../../utils/auth"
import { METHOD_GET, METHOD_OPTIONS, METHOD_POST, STATUS_200_OK } from "../../../utils/http"
import { ADMIN, STATS_GLOBAL } from "../../../utils/roles"

const handler = async (req, res) => {
  res.setHeader("Content-Type", "application/json")

  try {
    switch (req.method) {
      case METHOD_GET: {
        checkValidUserWithPrivilege(STATS_GLOBAL, req, res)

        const hospitals = await search(req.query)

        return res.status(STATUS_200_OK).json(hospitals)
      }
      case METHOD_POST: {
        const currentUser = checkValidUserWithPrivilege(ADMIN, req, res)

        checkIsSuperAdmin(currentUser)

        const id = await create(req.body)

        return res.status(STATUS_200_OK).json({ id })
      }
      default:
        if (req.method !== METHOD_OPTIONS) return sendMethodNotAllowedError(res)
    }
  } catch (error) {
    sendAPIError(error, res)
  }
}

const cors = Cors({
  allowMethods: [METHOD_GET, METHOD_OPTIONS],
})

export default cors(handler)
