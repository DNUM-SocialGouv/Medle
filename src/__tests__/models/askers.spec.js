import { transform, untransform } from "../../models/hospitals"

describe("transform", () => {
  it("should transform in a JS entity", () => {
    const knexData = {
      id: 3,
      name: "Brigade de gendarmerie de Saint-Aignan sur Cher",
      depCode: "41",
      type: "manual",
    }

    expect(transform(knexData)).toMatchSnapshot()
  })
})

describe("untransform", () => {
  it("should untransform in a DB entity", () => {
    const jsData = {
      id: 5,
      name: "Commissariat de police de Vincennes",
      depCode: "94",
      type: "manual",
    }

    expect(untransform(jsData)).toMatchSnapshot()
  })
})
