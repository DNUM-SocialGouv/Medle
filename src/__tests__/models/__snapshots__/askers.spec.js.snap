// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform should transform in a JS entity 1`] = `
Object {
  "addr1": undefined,
  "addr2": undefined,
  "depCode": undefined,
  "finesseNumber": undefined,
  "id": 3,
  "name": "Brigade de gendarmerie de Saint-Aignan sur Cher",
  "postalCode": undefined,
  "town": undefined,
}
`;

exports[`untransform should untransform in a DB entity 1`] = `
Object {
  "dep_code": "94",
  "extra_data": Object {
    "type": "manual",
  },
  "id": 5,
  "name": "Commissariat de police de Vincennes",
}
`;
